File: parser/rule/block/block.go
/*
   The Fluent Programming Language
   -----------------------------------------------------
   Copyright (c) 2025 Rodrigo R. & All Fluent Contributors
   This program comes with ABSOLUTELY NO WARRANTY.
   For details type `fluent l`. This is free software,
   and you are welcome to redistribute it under certain
   conditions; type `fluent l -f` for details.
*/

package block

import (
	"fluent/ast"
	"fluent/parser/error"
	"fluent/parser/queue"
	"fluent/parser/rule/conditional"
	declaration2 "fluent/parser/rule/declaration"
	"fluent/parser/rule/expression"
	"fluent/parser/rule/loop"
	"fluent/parser/rule/ret"
	"fluent/parser/util"
	"fluent/token"
)

// ProcessBlock processes a block of tokens and returns the corresponding AST and any parsing errors.
// Parameters:
// - input: a slice of tokens to be processed.
// Returns:
// - ast.AST: the abstract syntax tree representing the block.
// - error.Error: any error encountered during parsing.
func ProcessBlock(input []token.Token) (ast.AST, error.Error) {
	if len(input) == 0 {
		return ast.AST{
			Rule:     ast.Block,
			Children: &[]*ast.AST{},
		}, error.Error{}
	}

	// All tokens have been checked before they reach this function
	// Therefore it is not necessary to check the input length here
	result := ast.AST{
		Rule:     ast.Block,
		Line:     input[0].Line,
		Column:   input[0].Column,
		File:     &input[0].File,
		Children: &[]*ast.AST{},
	}

	// Use a queue for nested blocks
	var processQueue []queue.Element

	// Add the first element to the queue
	processQueue = append(processQueue, queue.Element{
		Tokens: input,
		Parent: &result,
	})

	for len(processQueue) > 0 {
		// Pop the first element from the queue
		element := processQueue[0]
		processQueue = processQueue[1:]

		// Subtract the data from the element
		input := element.Tokens
		result := element.Parent

		// Used to skip tokens
		skip := 0

		// Used for parsing ifs
		var lastIf *ast.AST

		for i, el := range input {
			if i < skip {
				continue
			}

			tokenType := el.TokenType

			// Handle conditionals
			switch tokenType {
			case token.If, token.OpenCurly, token.While, token.ElseIf, token.Else, token.For:
				// Extract the tokens in the nested block
				block := util.ExtractTokensBefore(
					input[i:],
					[]token.Type{token.CloseCurly},
					true,
					token.OpenCurly,
					token.CloseCurly,
					true,
				)

				if block == nil {
					return ast.AST{}, error.Error{
						Line:     el.Line,
						Column:   el.Column,
						File:     &el.File,
						Expected: []ast.Rule{ast.Block},
					}
				}

				skip = i + len(block) + 1

				if tokenType == token.OpenCurly {
					// Create a nested block node
					blockNode := ast.AST{
						Rule:     ast.Block,
						Line:     el.Line,
						Column:   el.Column,
						File:     &el.File,
						Children: &[]*ast.AST{},
					}

					// Add the nested block to the result
					*result.Children = append(*result.Children, &blockNode)

					// Schedule the nested block
					processQueue = append(processQueue, queue.Element{
						Tokens: block[1:], // Strip the open curly
						Parent: &blockNode,
					})
				} else {
					isIf := tokenType == token.If
					isElseIf := tokenType == token.ElseIf
					isElse := tokenType == token.Else
					isFor := tokenType == token.For

					if isElseIf || isElse {
						if lastIf == nil {
							return ast.AST{}, error.Error{
								Line:     el.Line,
								Column:   el.Column,
								File:     &el.File,
								Expected: []ast.Rule{ast.If},
							}
						}
					}

					// Parse the condition
					var condition ast.AST
					var parsingError error.Error

					if isFor {
						condition, parsingError = loop.ProcessForLoop(block, &processQueue)
					} else if isElse {
						condition, parsingError = conditional.ProcessElse(block, &processQueue)
					} else {
						condition, parsingError = conditional.ProcessLoopOrConditional(
							block,
							&processQueue,
							isIf,
							isElseIf,
						)
					}

					if parsingError.IsError() {
						return ast.AST{}, parsingError
					}

					if isIf {
						lastIf = &condition
					} else if isElseIf || isElse {
						*lastIf.Children = append(*lastIf.Children, &condition)
					}

					if !isElseIf && !isElse {
						*result.Children = append(*result.Children, &condition)
					}
				}
			default:
				// Extract a statement
				statement := util.ExtractTokensBefore(
					input[i:],
					[]token.Type{token.Semicolon},
					false,
					token.Unknown,
					token.Unknown,
					true,
				)

				if statement == nil {
					return ast.AST{}, error.Error{
						Line:     el.Line,
						Column:   el.Column,
						File:     &el.File,
						Expected: []ast.Rule{ast.Statement},
					}
				}

				skip = i + len(statement) + 1

				var specialCaseNode ast.AST
				var specialCaseError error.Error

				// Process special cases
				if tokenType == token.Let || tokenType == token.Const {
					// Process the declaration
					specialCaseNode, specialCaseError = declaration2.ProcessDeclaration(statement)
				} else if tokenType == token.Continue || tokenType == token.Break {
					// Process the continue or break statement
					specialCaseNode, specialCaseError = loop.ProcessContinueOrBreak(statement)
				} else if tokenType == token.Return {
					// Process the return statement
					specialCaseNode, specialCaseError = ret.ProcessReturn(statement)
				}

				// Check for special cases
				if specialCaseNode.Rule != ast.Program {
					if specialCaseError.IsError() {
						return ast.AST{}, specialCaseError
					}

					// Append the node to the result
					*result.Children = append(*result.Children, &specialCaseNode)

					// Avoid further processing
					continue
				}

				// Parse as an expression
				expr, parsingError := expression.ProcessExpression(statement)

				if parsingError.IsError() {
					return ast.AST{}, parsingError
				}

				*result.Children = append(*result.Children, &expr)
			}
		}
	}

	return result, error.Error{}
}

File: parser/rule/block/block.cpp
/*
        ==== The Zelix Programming Language ====
---------------------------------------------------------
  - This file is part of the Zelix Programming Language
    codebase. Zelix is a fast, statically-typed and
    memory-safe programming language that aims to
    match native speeds while staying highly performant.
---------------------------------------------------------
  - Zelix is categorized as free software; you can
    redistribute it and/or modify it under the terms of
    the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.
---------------------------------------------------------
  - Zelix is distributed in the hope that it will
    be useful, but WITHOUT ANY WARRANTY; without even
    the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU General Public
    License for more details.
---------------------------------------------------------
  - You should have received a copy of the GNU General
    Public License along with Zelix. If not, see
    <https://www.gnu.org/licenses/>.
*/

#include "block.h"
#include "parser/expect.h"
#include "parser/rule/assignment.h"
#include "parser/rule/declaration.h"
#include "parser/rule/expr/expr.h"
#include "parser/rule/for.h"
#include "parser/rule/if.h"
#include "parser/rule/ret.h"

using namespace zelix;

void parser::rule::block(
    ast *&root,
    container::stream<lexer::token *> &tokens,
    memory::lazy_allocator<ast> &allocator,
    const lexer::token *const &trace
)
{
    expect(tokens, lexer::token::OPEN_CURLY);
    tokens.next(); // Consume the open curly brace

    // Create a new block AST node
    ast *root_block = allocator.alloc();
    root_block->rule = ast::BLOCK;
    root->children.push_back(root_block);

    // Use a queue for the children of the block
    container::vector<ast *> block_queue;
    block_queue.push_back(root_block);

    // Save the current conditional for if/else blocks
    ast *current_conditional = nullptr;

    // Get the next token
    while (true)
    {
        auto next_opt = tokens.peek();
        if (next_opt.is_none() && !block_queue.empty())
        {
            global_err.type = UNEXPECTED_TOKEN;
            global_err.column = trace->column;
            global_err.line = trace->line;
            throw except::exception("Unexpected end of block");
        }

        // Get the current block node
        ast *current_block = block_queue.ref_at(block_queue.size() - 1);

        switch (
            const auto &next = next_opt.get();
            next->type
        )
        {
            case lexer::token::CLOSE_CURLY:
            {
                if (block_queue.empty())
                {
                    global_err.type = UNEXPECTED_TOKEN;
                    global_err.column = next->column;
                    global_err.line = next->line;
                    throw except::exception("Unexpected close curly brace without an open block");
                }

                block_queue.pop_back(); // Close the current block
                tokens.next(); // Consume the close curly brace
                if (block_queue.empty())
                {
                    return; // If the block queue is empty, we are done parsing the block
                }

                break;
            }

            case lexer::token::OPEN_CURLY:
            {
                // Create a new block node for the nested block
                ast *nested_block = allocator.alloc();
                nested_block->rule = ast::BLOCK;
                current_block->children.push_back(nested_block);
                tokens.next(); // Consume the open curly brace

                block_queue.push_back(nested_block);
                break;
            }

            // Handle variables
            case lexer::token::LET:
            {
                tokens.next(); // Consume the let token

                declaration<false>(
                    current_block,
                    tokens,
                    allocator
                );

                break;
            }

            case lexer::token::CONST:
            {
                tokens.next(); // Consume the const token

                declaration<true>(
                    current_block,
                    tokens,
                    allocator
                );

                break;
            }

            case lexer::token::IF:
            {
                tokens.next(); // Consume the token

                conditional<true, false, false, false>(
                    current_block,
                    current_conditional,
                    next,
                    tokens,
                    allocator
                );

                break;
            }

            case lexer::token::ELSEIF:
            {
                tokens.next(); // Consume the token

                conditional<false, true, false, false>(
                    current_block,
                    current_conditional,
                    next,
                    tokens,
                    allocator
                );

                break;
            }

            case lexer::token::ELSE:
            {
                tokens.next(); // Consume the token

                conditional<false, false, true, false>(
                    current_block,
                    current_conditional,
                    next,
                    tokens,
                    allocator
                );

                break;
            }

            case lexer::token::WHILE:
            {
                tokens.next(); // Consume the token

                conditional<false, false, false, true>(
                    current_block,
                    current_conditional,
                    next,
                    tokens,
                    allocator
                );
                break;
            }

            case lexer::token::FOR:
            {
                tokens.next(); // Consume the token

                for_loop(
                    current_block,
                    tokens,
                    allocator,
                    next
                );

                break;
            }

            case lexer::token::RETURN:
            {
                tokens.next(); // Consume the token
                ret(
                    tokens,
                    allocator,
                    next,
                    current_block
                );

                break;
            }

            case lexer::token::IDENTIFIER:
            {
                if (
                    assignment(
                        current_block,
                        tokens,
                        allocator,
                        next
                    )
                ) break; // Successfully parsed an assignment
            }

            default:
            {
                // Pass the expression to the expression parser
                expression(
                    current_block,
                    tokens,
                    allocator,
                    trace
                );
            }
        }
    }
}

